Add `estimate` parameter to `plot_raw_psd` function.

aweinstein opened the pull request:
This new parameter allows choosing between plotting the PSD and the ASD.
It also fixes an inconsistency of the `dB` parameter.

See issue #4642.

agramfort commented on mne/viz/raw.py:
```diff
@@ -566,31 +566,59 @@ def _set_psd_plot_params(info, proj, picks, ax, area_mode):
+        psds *= scaling
+        np.log10(psds, out=psds)
+        psds *= 10
+        ylabel = '$\\frac{%s}{\\sqrt{Hz}}\ (dB)$ ' % unit
```
I would simplify the branching structure.

First

```
if 'auto':
    if dB:
         estimate='power'
    else:
         estimate='amplitude'
```
then it's obvious what auto means
after:

```
if estimate == "amplitude":
     np.sqrt(psds, out=psds)
     psds *= scaling
     ylabel = '$\\frac{%s}{\\sqrt{Hz}}$ ' % unit
else:
     ...

if dB:
    ylabel += "$\ (dB)$
```

aweinstein replied:
OK. Just pushed a version with simplified control flow.

cbrnr commented on mne/viz/raw.py:
```diff
@@ -566,31 +566,59 @@ def _set_psd_plot_params(info, proj, picks, ax, area_mode):
+
+    where = np.where(psds.min(1) <= 0)[0]
+    if len(where) > 0:
+        raise ValueError("Infinite value in PSD for channel(s) %s. "
```
If `dB=False` the power is not infinite but zero.

aweinstein replied:
OK. I wonder if this should be a warning rather than an Exception (I imagine someone might have a channel short-circuited on purpose.

cbrnr replied:
+1 for warning

larsoner replied:
fine by me

aweinstein replied:
Throwing a warning opens a new can of worms. Now when computing the log10, I get NumPy warnings. Do we just let them there? They look kind of ugly. But disabling these warnings in the code is not nice either.

larsoner replied:
`np.log10(np.maximum(whatever, np.finfo(float).tiny))`

aweinstein replied:
Thanks. Just pushed with this suggestion.

cbrnr commented on mne/viz/raw.py:
```diff
@@ -566,31 +566,59 @@ def _set_psd_plot_params(info, proj, picks, ax, area_mode):
+    parameters `dB` and `estimate`, and the type of plot and corresponding
+    units.
+
+    | dB    | estimate    | plot | units           |
```
This table would be easier to read if you structured it differently:

```
| dB    | estimate    | plot | units            |
|-------+-------------+------+------------------|
| True  | 'power'     | PSD  | uV**2/Hz (dB)    |
| True  | 'amplitude' | ASD  | uV/sqrt(Hz) (dB) |
| True  | 'auto'      | PSD  | uV**2/Hz (dB)    |
| False | 'power'     | PSD  | uV**2/Hz         |
| False | 'amplitude' | ASD  | uV/sqrt(Hz)      |
| False | 'auto'      | ASD  | uV/sqrt(Hz)      |
```

Please also use `uV` and not `u` as units. Finally, I'd write `sqrt(Hz)` and not `sqrt{Hz}`.

aweinstein replied:
OK with the table order. Units are not uV though. I'm using u as a general unit, since this can be MEG, etc. I'll change it to 'amp', as in amplitude to avoid the confusion.

larsoner commented on mne/viz/raw.py:
```diff
@@ -637,7 +670,17 @@ def plot_raw_psd(raw, tmin=0., tmax=np.inf, fmin=0, fmax=np.inf, proj=False,
-        If True, transform data to decibels. If False, plot amplitudes.
+        Plot Power Spectral Density (PSD), in units (amplitude**2/Hz (dB)) if
+        `dB=True`, and `estimate='power'` or `estimate='auto'`. Plot PSD in
+        units (amplitude**2/Hz) if `dB=False` and, `estimate='power'. Plot
```
double-backticks on code:
```
... ``estimate='auto'`` ...
```

aweinstein replied:
Done.

larsoner commented on mne/viz/raw.py:
```diff
@@ -669,6 +712,7 @@ def plot_raw_psd(raw, tmin=0., tmax=np.inf, fmin=0, fmax=np.inf, proj=False,
     -------
     fig : instance of matplotlib figure
         Figure with frequency spectra of the data channels.
+
```
I don't think we need this and it might actually cause the style checkers some anguish

aweinstein replied:
Removed.

larsoner commented on mne/viz/raw.py:
```diff
@@ -566,31 +566,64 @@ def _set_psd_plot_params(info, proj, picks, ax, area_mode):
         ylabel = '$\\frac{%s}{\\sqrt{Hz}}$' % unit
+    else:
+        psds *= scaling * scaling
+        ylabel = '$%s^2/Hz$' % unit
```
Have you tried using `\textrm{}` on these to prevent italicizing?

aweinstein replied:
It seems \mathrm does the trick. Just pushed the change.

larsoner commented on mne/viz/raw.py:
```diff
@@ -566,31 +566,64 @@ def _set_psd_plot_params(info, proj, picks, ax, area_mode):
-        np.sqrt(psds, out=psds)
-        psds *= scaling
-        ylabel = '$\\frac{%s}{\\sqrt{Hz}}$' % unit
+        ylabel += '$\\ \\mathrm{(dB)}$'
```
Nice!

FYI no need to change it here (matter of taste) but in case you didn't know you can do `r'$\ \mathrm{(dB)}$` to avoid needing to double-backslash, which is often nice for these LaTeX expressions

aweinstein replied:
Looks nicer with the raw string. Just pushed the change.

aweinstein commented:
By the way, the following code checks all the combinations of dB and estimate:

https://gist.github.com/aweinstein/70803335da5414257a49ab1e54042eb7

larsoner commented:
PEP8

https://travis-ci.org/mne-tools/mne-python/jobs/286042459#L3054

larsoner commented:
You'll need to modify a test to use `warnings.catch_warnings(record=True) as w:` as we do elsewhere, instead of using `assert_raises`

https://travis-ci.org/mne-tools/mne-python/jobs/286295403#L2627

cbrnr commented on mne/viz/raw.py:
```diff
@@ -566,31 +566,64 @@ def _set_psd_plot_params(info, proj, picks, ax, area_mode):
+    dead_ch = ', '.join(ch_names[ii] for ii in where)
+    if len(where) > 0:
+        if dB:
+            msg = "Infinite value in PSD for channel(s) %s. " \
```
`str.format` should be preferred over `%` formatting. To avoid the long string, you could set a short string `msg = 'Zero'` or `msg = 'Infinite'` and then create the message after the condition.

larsoner replied:
> str.format should be preferred over % formatting.

FYI we actually don't generally follow this convention, but if it makes things simpler here go ahead

cbrnr replied:
OK, then it's probably fine as it is.

larsoner commented:
@agramfort okay with you to merge once CIs are happy?

larsoner commented:
Travis fails just because OSX fails to install properly. This should be good to go. (Incidentally, I get that same install error on my OSX machine, not sure why it's just now happened on Travis.)

@agramfort okay with you to merge?

@cbrnr you happy?

agramfort commented:
@larsoner no time to look. merge if you're happy

cbrnr commented:
The y-axis label in the PSD plots do not show a fraction but a simple slash, whereas the ASD plot use a fraction. This could be unified (I'm not sure what I like better, the fraction looks kind of tiny and is hard to read so I'd probably not use it). Furthermore, "Hz" is italic in the PSD plot (this should be \mathrm'd).

I'd not use divisions but negative exponentials in all cases, this looks cleaner to me, e.g. `$\mathrm{uV}^2 \mathrm{Hz}^{-1}\ \mathrm{(dB)}$` or `$\mathrm{uV} \mathrm{Hz}^{-\frac{1}{2}}\ \mathrm{(dB)}$` (untested LaTeX).

The x-axis label could be changed to "Frequency (Hz)" instead of "Freq (Hz)".

Probably not related to this PR (but maybe a quick fix), if `average=False` the y-axis label is yet another variant, namely "PSD (1/sqrt(Hz))". It would be nice if all PSD/ASD plots used consistent labels.

cbrnr commented:
But of course all of this could go into a separate PR, in which case I'm happy with this one.

agramfort commented:
+1 for mathrm abbreviations

larsoner commented:
> if average=False the y-axis label is yet another variant, namely "PSD (1/sqrt(Hz))"

How do you mean? The averaging over channels (or not) shouldn't change the units...?

> I'd not use divisions but negative exponentials in all cases, this looks cleaner to me

It's likely we use `/` elsewhere, so let's go with that for now for simplicity. We can think about negative exponentials in a follow-up if it looks better (would require some testing, discussion, and consensus)

larsoner commented:
> How do you mean?

Oh wait -- I think you mean the label changes when it shouldn't, in which case I agree it should stay the same.

cbrnr commented:
Yes, the unit is correct but the labels should be consistent.

aweinstein commented:
I updated the y label to use `/` instead of `\fraq`. Also corrected the x labe to `Frequency (Hz)`.

I don't understand why the x label changes when `average=False`. From the discussion above it is not clear to me if that's a bug, and it should remain the same, or if it should be adjusted depending on `dB` and `estimate`.

larsoner commented:
> I don't understand why the x label changes when average=False. From the discussion above it is not clear to me if that's a bug, and it should remain the same, or if it should be adjusted depending on dB and estimate.

I think it's a bug and should be fixed to behave like we've discussed here (which I assume works for `average=True`)

aweinstein commented:
I think I fixed it, although not in a very elegant way.

aweinstein commented:
Done.

larsoner commented:
Thanks @aweinstein
